go/chat/server_test.go

```diff
@@ -3631,3 +3632,133 @@ func TestChatSrvTeamChannelNameMentions(t *testing.T) {
+					require.Nil(t, searchHit.NextMessage)
+				} else {
+					require.True(t, searchHit.NextMessage.IsValid())
+					require.EqualValues(t, searchHit.NextMessage.Valid().MessageID, nextMessageID)
+				}
+
+			}
+			_verifyHit(searchHit)
+			select {
+			case searchHitArg := <-searchHitCb:
+				_verifyHit(searchHitArg.SearchHit)
+			case <-time.After(20 * time.Second):
+				require.Fail(t, "no search result received")
+			}
+		}
+		verifySearchDone := func(numHits int) {
+			select {
+			case searchDone := <-searchDoneCb:
+				require.Equal(t, searchDone.NumHits, numHits)
+			case <-time.After(20 * time.Second):
+				require.Fail(t, "no search result received")
+			}
+		}
+
+		search := func(query string, maxHits int, maxMessages int) chat1.GetSearchRegexpRes {
+			res, err := tc.chatLocalHandler().GetSearchRegexp(tc.startCtx, chat1.GetSearchRegexpArg{
+				ConversationID: conv.Id,
+				Query:          query,
+				MaxHits:        maxHits,
+				MaxMessages:    maxMessages,
+			})
+			require.NoError(t, err)
+			return res
+		}
+
+		maxHits := 1
+		maxMessages := 1000
+		// Test basic equality match
+		query := "hi"
+		msgBody := "hi"
+		messageID := sendMessage(msgBody)
+		res := search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), 1)
+		verifyHit(nullMessageID, messageID, nullMessageID, []string{query}, res.Hits[0])
+		verifySearchDone(1)
+
+		// Test basic no results
+		query = "hey"
+		res = search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), 0)
+		verifySearchDone(0)
+
+		// Test maxHits
+		maxHits = 1
+		query = "hi"
+		messageID1 := sendMessage(msgBody)
+		res = search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), 1)
+		verifyHit(messageID, messageID1, nullMessageID, []string{query}, res.Hits[0])
+		verifySearchDone(1)
+
+		maxHits = 5
+		res = search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), 2)
+		verifyHit(messageID, messageID1, nullMessageID, []string{query}, res.Hits[0])
+		verifyHit(nullMessageID, messageID, messageID1, []string{query}, res.Hits[1])
+		verifySearchDone(2)
+
+		messageID2 := sendMessage(msgBody)
+		res = search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), 3)
+		verifyHit(messageID1, messageID2, nullMessageID, []string{query}, res.Hits[0])
+		verifyHit(messageID, messageID1, messageID2, []string{query}, res.Hits[1])
+		verifyHit(nullMessageID, messageID, messageID1, []string{query}, res.Hits[2])
+		verifySearchDone(3)
+
+		// Test regex functionality
+		query = "h.*"
+		lowercase := "abcdefghijklmnopqrstuvwxyz"
+		for _, char := range lowercase {
+			sendMessage("h" + string(char))
+		}
+		maxHits = len(lowercase)
+		res = search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), maxHits)
+		verifySearchDone(maxHits)
+
+		// Test maxMessages
+		maxMessages = 2
+		res = search(query, maxHits, maxMessages)
+		require.Equal(t, len(res.Hits), maxMessages)
+		verifySearchDone(maxMessages)
+	})
```

Comment:
Please also test non-ASCII search with UTF-8 runes (ask @songgao for some fun pointers)

Reply:
Hmm did you mean Chinese samples or just how CJK works with chat CLI in general?

If former, Joshua=约书亚 and John=约翰. I'd also love to suggest 屌爆了 which in literal translate to "explosively badass", and we use it to describe something super awesome.

If latter, CJK characters have always worked but never perfectly in CLI because we didn't deal with character width and terminal gets confused, so CJK characters will end up in weird output shape.

Reply:
Yeah I meant Chinese samples. Anyway I just wanted tests with Chinese characters in messages and in the search, to make sure they work.

Reply:
OK here's a complete sentence, with a nice CJK punctuation character: 

> 约书亚和约翰屌爆了！

Reply:
done

Reply:
ugh github never refreshes properly anymore haha i'll add these!

Reply:
haha cool! FWIW I think what you have (⌘) is good enough since Go's `regexp` package explicitly says everything is UTF-8 encoded. But these are more fun so I'm OK with adding these :D